homeassistant/components/orangepi_gpio/binary_sensor.py

```diff
@@ -1,50 +1,57 @@
 """Support for binary sensor using Orange Pi GPIO."""
-import logging
 
-from homeassistant.components import orangepi_gpio
-from homeassistant.components.binary_sensor import BinarySensorDevice, PLATFORM_SCHEMA
-from homeassistant.const import DEVICE_DEFAULT_NAME
+from homeassistant.components.binary_sensor import (
+    ENTITY_ID_FORMAT,
+    PLATFORM_SCHEMA,
+    BinarySensorDevice,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.entity import async_generate_entity_id
 
-from . import CONF_PIN_MODE
-from .const import CONF_INVERT_LOGIC, CONF_PORTS, PORT_SCHEMA
-
-_LOGGER = logging.getLogger(__name__)
+from . import edge_detect, read_input, setup_input, setup_mode
+from .const import CONF_INVERT_LOGIC, CONF_PIN_MODE, CONF_PORTS, PORT_SCHEMA
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(PORT_SCHEMA)
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Orange Pi GPIO devices."""
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the Orange Pi GPIO platform."""
+    binary_sensors = []
+    invert_logic = config[CONF_INVERT_LOGIC]
     pin_mode = config[CONF_PIN_MODE]
-    orangepi_gpio.setup_mode(pin_mode)
+    ports = config[CONF_PORTS]
 
-    invert_logic = config[CONF_INVERT_LOGIC]
+    setup_mode(pin_mode)
 
-    binary_sensors = []
-    ports = config[CONF_PORTS]
     for port_num, port_name in ports.items():
-        binary_sensors.append(OPiGPIOBinarySensor(port_name, port_num, invert_logic))
-    add_entities(binary_sensors, True)
+        binary_sensors.append(
+            OPiGPIOBinarySensor(hass, port_name, port_num, invert_logic)
+        )
+    async_add_entities(binary_sensors)
 
 
 class OPiGPIOBinarySensor(BinarySensorDevice):
     """Represent a binary sensor that uses Orange Pi GPIO."""
 
-    def __init__(self, name, port, invert_logic):
+    def __init__(self, hass, name, port, invert_logic):
         """Initialize the Orange Pi binary sensor."""
-        self._name = name or DEVICE_DEFAULT_NAME
+        self.entity_id = async_generate_entity_id(ENTITY_ID_FORMAT, name, hass=hass)
+        self._name = name
         self._port = port
         self._invert_logic = invert_logic
         self._state = None
 
-        orangepi_gpio.setup_input(self._port)
+    async def async_added_to_hass(self):
+        """Run when entity about to be added to hass."""
 
-        def read_gpio(port):
-            """Read state from GPIO."""
-            self._state = orangepi_gpio.read_input(self._port)
-            self.schedule_update_ha_state()
+        @callback
```

Comment:
The callback will be called from another thread. We shouldn't decorate it with `@callback`, since it won't run in the event loop.